File: app/src/main/java/app/kamy/saatApp/domain/model/UserProfile.kt
package app.kamy.saatApp.domain.model

import kotlinx.serialization.Serializable

@Serializable
data class UserProfileAvatarUrls(
    val small: String? = null,
    val medium: String? = null,
    val large: String? = null
)

@Serializable
data class UserProfileSettings(
    val ayahLanguages: List<Int>? = null,
    val reflectionLanguages: List<Int>? = null
)

@Serializable
data class UserProfilePayload(
    val avatarUrls: UserProfileAvatarUrls? = null,
    val createdAt: String? = null,
    val joiningYear: Int? = null,
    val isPasswordSet: Boolean? = null,
    val settings: UserProfileSettings? = null,
    val username: String? = null,
    @Serializable(with = StringFromAnyScalarSerializer::class)
    val id: String,
    val verified: Boolean? = null,
    val postAs: Boolean? = null,
    val firstName: String? = null,
    val lastName: String? = null,
    val postsCount: Int? = null,
    val averageToxicity: Double? = null,
    val languageId: Int? = null,
    val banned: Boolean? = null,
    val memberType: Int? = null,
    val followersCount: Int? = null,
    val likesCount: Int? = null,
    val isAdmin: Boolean? = null,
    val languageIsoCode: String? = null,
    val bio: String? = null,
    val country: String? = null,
    val followed: Boolean? = null
) {
    val displayTitle: String
        get() {
            val parts = listOfNotNull(firstName, lastName).filter { it.isNotEmpty() }
            if (parts.isNotEmpty()) return parts.joinToString(" ")
            if (!username.isNullOrEmpty()) return username
            return if (id.isEmpty()) "Profile" else id
        }

    val preferredAvatarUrl: String?
        get() = avatarUrls?.medium ?: avatarUrls?.large ?: avatarUrls?.small
}

File: iosApp/Saat/Domain/Models/UserProfile.swift
//
//  UserProfile.swift
//  Saat
//
//  Created by Elmee on 25/04/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

struct UserProfileAvatarUrls: Decodable, Sendable {
    let small: String?
    let medium: String?
    let large: String?
}

struct UserProfileSettings: Decodable, Sendable {
    let ayahLanguages: [Int]?
    let reflectionLanguages: [Int]?
}

struct UserProfilePayload: Decodable, Sendable {
    let avatarUrls: UserProfileAvatarUrls?
    let createdAt: String?
    let joiningYear: Int?
    let isPasswordSet: Bool?
    let settings: UserProfileSettings?
    let username: String?
    let id: String
    let verified: Bool?
    let postAs: Bool?
    let firstName: String?
    let lastName: String?
    let postsCount: Int?
    let averageToxicity: Double?
    let languageId: Int?
    let banned: Bool?
    let memberType: Int?
    let followersCount: Int?
    let likesCount: Int?
    let isAdmin: Bool?
    let languageIsoCode: String?
    let bio: String?
    let country: String?
    let followed: Bool?

    enum CodingKeys: String, CodingKey {
        case avatarUrls, createdAt, joiningYear, isPasswordSet, settings
        case username, id, verified, postAs, firstName, lastName
        case postsCount, averageToxicity, languageId, banned, memberType
        case followersCount, likesCount, isAdmin, languageIsoCode, bio, country, followed
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        avatarUrls = try c.decodeIfPresent(UserProfileAvatarUrls.self, forKey: .avatarUrls)
        createdAt = try c.decodeIfPresent(String.self, forKey: .createdAt)
        joiningYear = try c.decodeIfPresent(Int.self, forKey: .joiningYear)
        isPasswordSet = try c.decodeIfPresent(Bool.self, forKey: .isPasswordSet)
        settings = try c.decodeIfPresent(UserProfileSettings.self, forKey: .settings)
        username = try c.decodeIfPresent(String.self, forKey: .username)

        if let s = try? c.decode(String.self, forKey: .id) {
            id = s
        } else if let i = try? c.decode(Int.self, forKey: .id) {
            id = String(i)
        } else {
            id = ""
        }

        verified = try c.decodeIfPresent(Bool.self, forKey: .verified)
        postAs = try c.decodeIfPresent(Bool.self, forKey: .postAs)
        firstName = try c.decodeIfPresent(String.self, forKey: .firstName)
        lastName = try c.decodeIfPresent(String.self, forKey: .lastName)
        postsCount = try c.decodeIfPresent(Int.self, forKey: .postsCount)

        if let d = try? c.decode(Double.self, forKey: .averageToxicity) {
            averageToxicity = d
        } else if let i = try? c.decode(Int.self, forKey: .averageToxicity) {
            averageToxicity = Double(i)
        } else {
            averageToxicity = nil
        }

        languageId = try c.decodeIfPresent(Int.self, forKey: .languageId)
        banned = try c.decodeIfPresent(Bool.self, forKey: .banned)
        memberType = try c.decodeIfPresent(Int.self, forKey: .memberType)
        followersCount = try c.decodeIfPresent(Int.self, forKey: .followersCount)
        likesCount = try c.decodeIfPresent(Int.self, forKey: .likesCount)
        isAdmin = try c.decodeIfPresent(Bool.self, forKey: .isAdmin)
        languageIsoCode = try c.decodeIfPresent(String.self, forKey: .languageIsoCode)
        bio = try c.decodeIfPresent(String.self, forKey: .bio)
        country = try c.decodeIfPresent(String.self, forKey: .country)
        followed = try c.decodeIfPresent(Bool.self, forKey: .followed)
    }

    var displayTitle: String {
        let parts = [firstName, lastName].compactMap { $0 }.filter { !$0.isEmpty }
        if parts.isEmpty == false { return parts.joined(separator: " ") }
        if let u = username, u.isEmpty == false { return u }
        return id.isEmpty ? "Profile" : id
    }

    var preferredAvatarURL: URL? {
        let s = avatarUrls?.medium ?? avatarUrls?.large ?? avatarUrls?.small
        guard let s, let u = URL(string: s) else { return nil }
        return u
    }
}
